File: packages/telemetry/metrics.go
package telemetry

import (
	"context"
	"sync"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

const MeterName = "sentinel"

var (
	counterMu sync.Mutex
	counters  = map[string]metric.Int64Counter{}
	histos    = map[string]metric.Float64Histogram{}
	updowns   = map[string]metric.Int64UpDownCounter{}
	gauges    = map[string]metric.Float64Gauge{}
)

func Attrs(kv ...string) []attribute.KeyValue {
	out := make([]attribute.KeyValue, 0, 2+len(kv)/2)
	out = append(out,
		attribute.String("service", ServiceName()),
		attribute.String("environment", Environment()),
	)
	for i := 0; i+1 < len(kv); i += 2 {
		out = append(out, attribute.String(kv[i], kv[i+1]))
	}
	return out
}

func Count(ctx context.Context, name string, kv ...string) {
	c := counter(name)
	c.Add(ctx, 1, metric.WithAttributes(Attrs(kv...)...))
}

func Add(ctx context.Context, name string, n int64, kv ...string) {
	c := counter(name)
	c.Add(ctx, n, metric.WithAttributes(Attrs(kv...)...))
}

func Observe(ctx context.Context, name string, seconds float64, kv ...string) {
	h := histogram(name)
	h.Record(ctx, seconds, metric.WithAttributes(Attrs(kv...)...))
}

func IncrActive(ctx context.Context, name string, delta int64, kv ...string) {
	u := updown(name)
	u.Add(ctx, delta, metric.WithAttributes(Attrs(kv...)...))
}

func Gauge(ctx context.Context, name string, value float64, kv ...string) {
	g := gauge(name)
	g.Record(ctx, value, metric.WithAttributes(Attrs(kv...)...))
}

func counter(name string) metric.Int64Counter {
	counterMu.Lock()
	defer counterMu.Unlock()
	if c, ok := counters[name]; ok {
		return c
	}
	c, _ := otel.Meter(MeterName).Int64Counter(name)
	counters[name] = c
	return c
}

func histogram(name string) metric.Float64Histogram {
	counterMu.Lock()
	defer counterMu.Unlock()
	if h, ok := histos[name]; ok {
		return h
	}
	h, _ := otel.Meter(MeterName).Float64Histogram(name, metric.WithUnit("s"), metric.WithExplicitBucketBoundaries(
		0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10,
	))
	histos[name] = h
	return h
}

func updown(name string) metric.Int64UpDownCounter {
	counterMu.Lock()
	defer counterMu.Unlock()
	if u, ok := updowns[name]; ok {
		return u
	}
	u, _ := otel.Meter(MeterName).Int64UpDownCounter(name)
	updowns[name] = u
	return u
}

func gauge(name string) metric.Float64Gauge {
	counterMu.Lock()
	defer counterMu.Unlock()
	if g, ok := gauges[name]; ok {
		return g
	}
	g, _ := otel.Meter(MeterName).Float64Gauge(name)
	gauges[name] = g
	return g
}

const (
	HTTPRequests        = "sentinel.http.requests"
	HTTPDuration        = "sentinel.http.request.duration"
	HTTPActive          = "sentinel.http.requests.active"
	KafkaPublished      = "sentinel.kafka.messages_published"
	KafkaConsumed       = "sentinel.kafka.messages_consumed"
	KafkaFailures       = "sentinel.kafka.processing_failures"
	KafkaLag            = "sentinel.kafka.consumer.lag"
	KafkaDuration       = "sentinel.kafka.processing.duration"
	DetectionEvents     = "sentinel.detection.events_processed"
	DetectionCreated    = "sentinel.detection.alerts_created"
	DetectionSuppressed = "sentinel.detection.alerts_suppressed"
	DetectionDuration   = "sentinel.detection.duration"
	DetectionRules      = "sentinel.detection.rule_triggers"
	IncidentCorrelated  = "sentinel.incident.alerts_correlated"
	IncidentCreated     = "sentinel.incident.incidents_created"
	IncidentAttached    = "sentinel.incident.alerts_attached"
	IncidentFailures    = "sentinel.incident.correlation_failures"
	IncidentDuration    = "sentinel.incident.duration"
	AIStarted           = "sentinel.ai.investigations_started"
	AICompleted         = "sentinel.ai.investigations_completed"
	AIFailed            = "sentinel.ai.investigations_failed"
	AIDuration          = "sentinel.ai.investigation.duration"
	AIToolCalls         = "sentinel.ai.tool_calls"
	AIToolFailures      = "sentinel.ai.tool_failures"
	AIModelRequests     = "sentinel.ai.model_requests"
	AIModelFailures     = "sentinel.ai.model_failures"
	AITokens            = "sentinel.ai.tokens"
	RemRequested        = "sentinel.remediation.requested"
	RemApproved         = "sentinel.remediation.approved"
	RemRejected         = "sentinel.remediation.rejected"
	RemStarted          = "sentinel.remediation.started"
	RemSucceeded        = "sentinel.remediation.succeeded"
	RemFailed           = "sentinel.remediation.failed"
	RemVerifyPassed     = "sentinel.remediation.verification_passed"
	RemVerifyFailed     = "sentinel.remediation.verification_failed"
	RemDuration         = "sentinel.remediation.duration"
	DBDuration          = "sentinel.db.operation.duration"
	DBErrors            = "sentinel.db.errors"
	DBPoolAcquired      = "sentinel.db.pool.acquired"
	IncidentActive      = "sentinel.incident.active"
)

func RecordPool(ctx context.Context, acquired float64) {
	Gauge(ctx, DBPoolAcquired, acquired, "operation", "pool")
}

File: services/ai/app/observability/metrics.py
from __future__ import annotations

from prometheus_client import Counter, Gauge, Histogram

SERVICE = "sentinel-ai"

HTTP_REQUESTS = Counter(
    "sentinel_http_requests",
    "HTTP requests",
    ["service", "environment", "method", "route", "status"],
)
HTTP_DURATION = Histogram(
    "sentinel_http_request_duration_seconds",
    "HTTP request duration",
    ["service", "environment", "method", "route", "status"],
)
HTTP_ACTIVE = Gauge(
    "sentinel_http_requests_active",
    "In-flight HTTP requests",
    ["service", "environment", "method", "route"],
)
KAFKA_PUBLISHED = Counter(
    "sentinel_kafka_messages_published",
    "Kafka publishes",
    ["service", "environment", "topic", "operation"],
)
KAFKA_CONSUMED = Counter(
    "sentinel_kafka_messages_consumed",
    "Kafka consumes",
    ["service", "environment", "topic", "operation"],
)
KAFKA_FAILURES = Counter(
    "sentinel_kafka_processing_failures",
    "Kafka processing failures",
    ["service", "environment", "operation", "topic"],
)
AI_STARTED = Counter(
    "sentinel_ai_investigations_started",
    "Investigations started",
    ["service", "environment", "status"],
)
AI_COMPLETED = Counter(
    "sentinel_ai_investigations_completed",
    "Investigations completed",
    ["service", "environment", "status"],
)
AI_FAILED = Counter(
    "sentinel_ai_investigations_failed",
    "Investigations failed",
    ["service", "environment", "status"],
)
AI_DURATION = Histogram(
    "sentinel_ai_investigation_duration_seconds",
    "Investigation duration",
    ["service", "environment", "status"],
)
AI_TOOL_CALLS = Counter(
    "sentinel_ai_tool_calls",
    "Tool calls",
    ["service", "environment", "tool", "status"],
)
AI_TOOL_FAILURES = Counter(
    "sentinel_ai_tool_failures",
    "Tool failures",
    ["service", "environment", "tool", "status"],
)
AI_MODEL_REQUESTS = Counter(
    "sentinel_ai_model_requests",
    "Model requests",
    ["service", "environment", "provider"],
)
AI_MODEL_FAILURES = Counter(
    "sentinel_ai_model_failures",
    "Model failures",
    ["service", "environment", "provider"],
)
AI_TOKENS = Counter(
    "sentinel_ai_tokens",
    "Model tokens when the provider reports usage",
    ["service", "environment", "kind"],
)
DB_DURATION = Histogram(
    "sentinel_db_operation_duration_seconds",
    "Database operation duration",
    ["service", "environment", "operation"],
)
DB_ERRORS = Counter(
    "sentinel_db_errors",
    "Database errors",
    ["service", "environment", "operation"],
)
